File: backend/document/schemas/document.py
"""Document Domain Pydantic DTOs & Schemas.

Includes the canonical Document Manifest DTO, response schemas, and detail views.
"""

from datetime import datetime
from typing import Any
import uuid

from pydantic import BaseModel, ConfigDict, Field


class StageMetricDTO(BaseModel):
    """Processing stage duration and execution details."""

    stage: str = Field(description="Stage name (e.g. validation, storage, extraction, ocr)")
    started_at: str | None = None
    completed_at: str | None = None
    duration_ms: float = Field(default=0.0, description="Duration in milliseconds")
    status: str = Field(default="COMPLETED")
    error_message: str | None = None


class DocumentManifestDTO(BaseModel):
    """Canonical Document Manifest generated upon successful pipeline execution."""

    manifest_version: str = Field(default="1.0.0", description="Manifest schema version")
    document_id: uuid.UUID
    version_id: uuid.UUID
    version_number: int
    tenant_id: str
    owner_user_id: uuid.UUID | None
    filename: str
    original_filename: str
    mime_type: str
    file_size_bytes: int
    checksum_sha256: str
    storage_provider: str
    original_storage_key: str
    normalized_text_path: str | None = None
    metadata_json_path: str | None = None
    page_count: int = 0
    word_count: int = 0
    language: str | None = None
    encoding: str | None = None
    stage_metrics: list[StageMetricDTO] = Field(default_factory=list)
    extraction_metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str


class DocumentVersionDTO(BaseModel):
    """Version detail summary schema."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    document_id: uuid.UUID
    version_number: int
    storage_object_id: uuid.UUID
    content_hash: str
    extracted_text_path: str | None = None
    metadata_json: dict[str, Any] | None = None
    created_at: datetime


class DocumentResponse(BaseModel):
    """Standard summary response representation of a Document aggregate."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    tenant_id: str
    owner_user_id: uuid.UUID | None = None
    filename: str
    original_filename: str
    status: str
    latest_version_id: uuid.UUID | None = None
    word_count: int = 0
    page_count: int = 0
    language: str | None = None
    created_at: datetime
    updated_at: datetime


class DocumentDetailResponse(DocumentResponse):
    """Extended document representation including version history and canonical manifest."""

    versions: list[DocumentVersionDTO] = Field(default_factory=list)
    manifest: DocumentManifestDTO | None = None


class DocumentListResponse(BaseModel):
    """Paginated document list response."""

    items: list[DocumentResponse]
    total: int
    page: int
    page_size: int
    pages: int

File: frontend/src/types/document.ts
/**
 * Document Intelligence frontend domain models (`ADR-005`).
 */

export interface StageMetricDTO {
  stage: string
  duration_ms: number
  status: string
}

export interface DocumentManifestDTO {
  manifest_version: string
  document_id: string
  version_id: string
  version_number: number
  tenant_id: string
  owner_user_id?: string | null
  filename: string
  original_filename: string
  mime_type: string
  file_size_bytes: number
  checksum_sha256: string
  storage_provider: string
  original_storage_key: string
  normalized_text_path?: string | null
  metadata_json_path?: string | null
  page_count: number
  word_count: number
  language?: string
  encoding?: string | null
  stage_metrics: StageMetricDTO[]
  extraction_metadata: Record<string, unknown>
  created_at: string
}

export interface DocumentVersionDTO {
  id: string
  document_id: string
  version_number: number
  storage_object_id: string
  content_hash: string
  extracted_text_path?: string
  metadata_json?: Record<string, unknown>
  created_at: string
}

export interface DocumentResponse {
  id: string
  tenant_id: string
  filename: string
  original_filename: string
  status: string
  latest_version_id?: string
  word_count: number
  page_count: number
  language?: string
  created_at: string
  updated_at: string
}

export interface DocumentDetailResponse extends DocumentResponse {
  versions: DocumentVersionDTO[]
  manifest?: DocumentManifestDTO
}

export interface ProcessingStatusResponse {
  document_id: string
  status: string
  current_step: string
  progress_percent: number
  retry_count: number
  error_code?: string
  error_message?: string
  updated_at: string
}

export interface UploadResponse {
  document_id: string
  version_id: string
  job_id: string
  status: string
  filename: string
  original_filename: string
  file_size_bytes: number
  created_at: string
}

export interface DocumentListResponse {
  items: DocumentResponse[]
  total: number
  page: number
  page_size: number
  pages: number
}
